os/services/shell/shell.c

```diff
@@ -89,7 +89,7 @@ output_prompt(shell_output_func output)
 PT_THREAD(shell_input(struct pt *pt, shell_output_func output, const char *cmd))
 {
   static char *args;
-  static struct shell_command_t *cmd_ptr;
+  const struct shell_command_t *cmd_descr = NULL;
```

Comment:
The `cmd_descr` variable is used by a protothread - it needs to be `static`.

Reply:
It seems this causes the test failures; I've not understood what really happened in the tests of `09-ipv6`, though.
```
$ git log --oneline | head -n1
0e17ee444 Introduce dynamically registered shell command sets

$ git diff --ignore-submodules
diff --git a/os/services/shell/shell.c b/os/services/shell/shell.c
index 359038f3f..c219b1903 100644
--- a/os/services/shell/shell.c
+++ b/os/services/shell/shell.c
@@ -89,7 +89,7 @@ output_prompt(shell_output_func output)
 PT_THREAD(shell_input(struct pt *pt, shell_output_func output, const char *cmd))
 {
   static char *args;
-  const struct shell_command_t *cmd_descr = NULL;
+  static const struct shell_command_t *cmd_descr = NULL;

   PT_BEGIN(pt);


$ contiker make -C tests/09-ipv6
make: Entering directory '/home/user/contiki-ng/tests/09-ipv6'
Running test 01-ping-lla-csma-w-rpl with random Seed 1................................................ OK
Running test 02-ping-ula-csma-w-rpl with random Seed 1........................................ OK
Running test 03-ping-lla-tsch-w-rpl with random Seed 1................................................... OK
Running test 04-ping-ula-tsch-w-rpl with random Seed 1............................................. OK
Running test 05-ping-lla-csma-wo-rpl with random Seed 1.................................. OK
Running test 06-ping-ula-csma-wo-rpl with random Seed 1.................................. OK
Running test 07-ping-lla-tsch-wo-rpl with random Seed 1.................................... OK
Running test 08-ping-ula-tsch-wo-rpl with random Seed 1................................... OK
========== Summary ==========
01-ping-lla-csma-w-rpl                   TEST OK      1/1
02-ping-ula-csma-w-rpl                   TEST OK      1/1
03-ping-lla-tsch-w-rpl                   TEST OK      1/1
04-ping-ula-tsch-w-rpl                   TEST OK      1/1
05-ping-lla-csma-wo-rpl                  TEST OK      1/1
06-ping-ula-csma-wo-rpl                  TEST OK      1/1
07-ping-lla-tsch-wo-rpl                  TEST OK      1/1
08-ping-ula-tsch-wo-rpl                  TEST OK      1/1
make: Leaving directory '/home/user/contiki-ng/tests/09-ipv6'
```

Reply:
OK; here is what happened in `01-ping-lla-csma-w-rpl`:

1. `ping` command is issued on node-1
1. `cmd_ping` waits at [L261](https://github.com/angelos-oikonomopoulos/contiki-ng/blob/0e17ee444740420e1d072426612c26cc3cc7a3b6/os/services/shell/shell-commands.c#L261) of of `shell-command.c` 
1. node-1 receives a ping response
1. the event starts being delivered to `cmd_ping` via `shell_input`
1. `shell_input` resumes its job at [L111](https://github.com/angelos-oikonomopoulos/contiki-ng/blob/0e17ee444740420e1d072426612c26cc3cc7a3b6/os/services/shell/shell.c#L111) of `shell.c`

At this moment, `cmd_descr` is `NULL` because of missing `static`. The null pointer reference causes `SIGSEGV`. The same should have happened in the other tests.

Reply:
The 09-ipv6 ping test is calling the `ping` shell command which is an asynchronous command. After the initial setup of the ping command, the command protothread will be called again later (either due to ping response or timeout) but at that time `cmd_desc` contains whatever happens to be on the stack instead of the address to the command protothread. Making `cmd_desc` static should solve the problem.

Reply:
Right, `cmd_desc` is set to `NULL` before `PT_BEGIN()` so it will always be `NULL` instead of whatever was on the stack when the protothread next executed.

Reply:
Thanks for tracking this down!

@yatch re: `4. the event starts being delivered to cmd_ping via shell_input` I don't think I'm following. How does the ping response involve `shell_input`?

AFAIU, the difference here is that `cmd_ping` is asynchronous, so `PT_SPAWN` will end up scheduling it again (i.e. calling it through `cmd_descr`, which will indeed be `NULL`  the second time we enter `shell_input`).

So, IIUC, the reason that I didn't observe that is I only tested with synchronous commands.

The main reason I'm asking is to understand if I should also use static in `cmd_help`. This depends on the semantics of the `shell_output_func`. Is it supposed to be synchronous across all platforms and configurations?

Reply:
@angelos-oikonomopoulos you don't need to make the variable `static` in `cmd_help`.

A simple rule about this is, if you wait for something in a `PT_THREAD`, you have a high chance to need make variables `static` there (of course, it depends on the context, though).  Otherwise, you don't need to do and you can save a little bit of static area of the memory.

Regarding `shell_input`, it looks like it doesn't have any wait in it; but in fact, `PT_WAIT_THREAD` is called as part of `PT_SPAWN`. 

It all happens under the magic of `pt` and `lc`. These implementation may be interesting to you.